### ISSUE TITLE
Lint/UselessAssignment False Positive

### ISSUE THREAD

**Issue by jamie** [rubocop/rubocop#13990]
I've found a situation where Lint/UselessAssignment will erroneously flag a variable as unused, and the autocorrect will cause functional Ruby to start erroring.

--------

## Expected behavior

Rubocop should not flag the assignment at line 5 (see below) as unused, as it is used at the end of the method, line 16.

## Actual behavior

```
❯ rubocop --debug useless_assignment.rb
For /Users/jamiemacey/code/rubo: Default configuration from /Users/jamiemacey/code/rubo/.mise/ruby/gems/rubocop-1.74.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/jamiemacey/code/rubo/useless_assignment.rb
W

Offenses:

useless_assignment.rb:3:1: C: Metrics/MethodLength: Method has too many lines. [12/10]
def translate(values) ...
^^^^^^^^^^^^^^^^^^^^^
useless_assignment.rb:5:5: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - output.
    output = values.map(&:to_s)
    ^^^^^^

1 file inspected, 2 offenses detected, 1 offense autocorrectable
Finished in 0.17228400008752942 seconds
```

## Steps to reproduce the problem

This is a reduced example from actual application code:

```
# frozen_string_literal: false

def translate(values)
  if values.first == 1
    output = values.map(&:to_s)
  else
    accumulated = ''
    output = []
    values.each do |value|
      accumulated << value.to_s
      output = [accumulated]
    end
    output
  end

  output.compact
end

pp translate([1, 3])
pp translate([2, 3, 4])
pp translate([])
```

I've attempted to simplify further, but a few attempts to get rid of `accumulated` or to hoist the `output = [accumulated]` outside the iterator line 9 prevent the bug from surfacing.

## RuboCop version

With no .rubocop.yml config file:

```
❯ ruby -v
ruby 3.4.2 (2025-02-15 revision d2930f8e7a) +PRISM [arm64-darwin24]

❯ rubocop -V
1.74.0 (using Parser 3.3.7.1, rubocop-ast 1.38.1, analyzing as Ruby 2.7, running on ruby 3.4.2) [arm64-darwin24]
```

Behaviour persists if I target Ruby 3.4:

```
❯ cat .rubocop.yml 
AllCops:
  TargetRubyVersion: 3.4
  NewCops: enable

❯ rubocop -V
1.74.0 (using Parser 3.3.7.1, rubocop-ast 1.38.1, analyzing as Ruby 3.4, running on ruby 3.4.2) [arm64-darwin24]

❯ rubocop --debug useless_assignment.rb
For /Users/jamiemacey/code/rubo: configuration from /Users/jamiemacey/code/rubo/.rubocop.yml
Default configuration from /Users/jamiemacey/code/rubo/.mise/ruby/gems/rubocop-1.74.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning /Users/jamiemacey/code/rubo/useless_assignment.rb
W

Offenses:

useless_assignment.rb:3:1: C: Metrics/MethodLength: Method has too many lines. [12/10]
def translate(values) ...
^^^^^^^^^^^^^^^^^^^^^
useless_assignment.rb:5:5: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - output.
    output = values.map(&:to_s)
    ^^^^^^

1 file inspected, 2 offenses detected, 1 offense autocorrectable
Finished in 0.18073100002948195 seconds
```
